### DISCUSSION TITLE
sole relationship modifier

### DISCUSSION THREAD

**Discussion by shaedrich** [laravel/framework#49913] [Ideas]
There's already `Relation::sole()`, but it isn't a modifier but executes the query, so it can't be used with `with()`, for example.

Let's assume, we have a process with multiple steps. Within the first step, a relationship might return `null`, but will be set by the end of said step.

```php
class ChildModel
{
    public function parent(): HasOne
    {
        return $this->hasOne(ParentModel::class);
    }
}
```

In the subsequent steps, I want to eager load said model but there might be cases, where we get an exception like this:

> [!CAUTION]
> Attempt to read property "name" on null

If I don't want to check this manually, maybe even in several separate places and don't want an accessor in the model, I can have a separate relationship:

```php
class ChildModel
{
    // parent relationship

    public function mandatoryParent()
    {
        return $this->parent()->withDefault(fn (Model $relation) => throw (new ModelNotFoundException)->setModel($relation::class));
    }
}
```

This works, however, it would be even nicer, if there was an out-of-the-box modifier to achive this without manually instanciating the exception and setting the model like with `sole()`.

**Comment by ghost** [comment]
I guess you are mentioning this because you are using the relation with magic call like 
$child->parent->name
What is wrong with 
$child->parent?->name
?

**Reply by shaedrich** [reply]
I do. Sometimes, you want `null`, but sometimes, you want an exception because neither `parent` nor `parent.name` can be `null`

**Reply by ghost** [reply]
\Illuminate\Database\Eloquent\Concerns\HasAttributes::getRelationValue

   /**
     * Get a relationship.
     *
     * @param  string  $key
     * @return mixed
     */
    public function getRelationValue($key)
    {
        // If the key already exists in the relationships array, it just means the
        // relationship has already been loaded, so we'll just return it out of
        // here because there is no need to query within the relations twice.
        if ($this->relationLoaded($key)) {
            return $this->relations[$key];
        }

You can overwrite this method as

    $parent = parent::...'

    if ($parent === null) { // is this null also for one to many relation?
     throw
    } 
      return $parent;

in your model

**Reply by shaedrich** [reply]
as I said:
> [&hellip;] I don't want to check this manually, maybe even in several separate places and don't want an accessor in the model [&hellip;] [but] a separate relationship.
